### ISSUE TITLE
MarkAsJunk email-learn issue with Virtualmin

### ISSUE THREAD

**Issue by diegoweb** [roundcube/roundcubemail#7489]
Hey guys!

I'm trying to set email-learn to work with virtualmin + roundcube in the MarkAsJunk plugin, but there's no option to capture the domain from the user actual email.

Since Virtualmin allow spamtrap and hamtrap aliases to be created, it's easier to report mails to those address.
Unfortunately there's no option available to fulfill this.

The closest one is **%i is replaced with the email address from the user's default identity**, but this will get the username from the user as well.

Basically what I need is to send spams to:
spamtrap@userdomain.tld

and ham mails to:
hamtrap@userdomain.tld

By inserting this into the debug section of the the email_learn.php driver:
`rcube::write_log('markasjunk', ' Session ' . $_SESSION['username']. ' Local ' . $this->rcube->user->get_username('local'). ' Domain ' . $this->rcube->user->get_username('domain'). ' From ' . $from);`

I was able to detect this in the markasjunk.log file:
`Session testuser Local testuser Domain mail.mailserver.tld From testuser@userdomain.tld`

So if I change the this section, I can work with the new variable called %v (this is just for test purpose, I don't even know if this would be the correct way to fix this issue, but it worked, this is what is important to me at this moment 🤣 ):

```
    private function _parse_vars($data, $spam, $from)
    {
	$virtualmin = explode('@',$from);
        $data = str_replace('%u', $_SESSION['username'], $data);
        $data = str_replace('%t', $spam ? 'spam' : 'ham', $data);
        $data = str_replace('%l', $this->rcube->user->get_username('local'), $data);
        $data = str_replace('%d', $this->rcube->user->get_username('domain'), $data);
        $data = str_replace('%i', $from, $data);
	$data = str_replace('%v', $virtualmin[1], $data);

        return $data;
    }
```

So in my config.inc.php I could simply set to:
```
$config['markasjunk_email_spam'] = 'spamtrap@%v';
$config['markasjunk_email_ham'] = 'hamtrap@%v';
```

Can I request this to be in the new roundcube versions? Not the way I did because probably there's a better way to do, but I think you guys understood what Virtualmin is looking for.

Thanks!

**Comment by johndoh** [comment]
Your user ids are in the format testuser@mail.mailserver.tld? Are you sure you do not have a general issue with your Roundcube config and actually what you want to do is set `mail_domain` in your main Roundcube config file?

**Comment by diegoweb** [comment]
`mail_domain` was not set.
I've set it in the config.inc.php from Roundcube:
`$config['mail_domain'] = '%t';`
I've tried all variables available for this config, but none of them is working as expected.

Basically I have one main domain that I host my roundcube. Since Postfix doesn't allow multiple SSL Certificates, I had to only have 1 domain for SMTP login. So all domains in my system log into the same IMAP Domain:
`mail.mysystem.com`

So both `default_host` and `smtp_server`  have `mail.mysystem.com` listed.

So based in my setup, what would be the correct input for `mail_domain` ? I can't rely on user imap, since it will always be my system domain. I need to use the domain part from the user login, which they use in this pattern: username@domain.tld.

This is what I'm getting with %d in the `markasjunk_email_spam`:

> [16/Jul/2020 22:25:48]: <eb7i1nbk>  Session testuser-userdomain.com.br Local testuser-userdomain.com.br Domain com.br From testuser@userdomain.com.br

So basically if I set %l in `markasjunk_email_spam` I get spamtrap@testuser-userdomain.com.br.
Even for the Password plugin, in Virtualmin driver, I had to hack some of the code to change from `-` to `@`
In Virtualmin the default is `@`, so I can't understand why Roundcube still insists to use `-` as a "domain splitter".

Even in this Roundcube URL: webmin/?_task=settings&_action=identities
The default name for every user is username-domain.tld

**Comment by johndoh** [comment]
so your username format is `local-domain.tld` not `local@domain.tld`? Or may be you have some other plugin which is messing with the username?

If the user logged in with `username@domain.tld` like you say then using `spamtrap@%d` in `markasjunk_email_spam` should resolve just fine but you say that `$_SESSION['username']` comes back as `testuser-userdomain.com.br`, how did the `@` change to a `-`? Roundcube does not do that.

A macro to use the domain part from the primary identity could be added but I still feel like there is something else going on here. Also just for my own curiosity, you have a shared mail system right? so what is so important about the spamtrap@ address, could you not send the emails to a central address which work for all domains? Do you not use the sender information to know who is reporting the spam?

@alecpl I think I have seen tickets in the past about using something other than @ between the local and domain parts in usernames but I cannot find anything relevant now. Do you remember anything like that?

**Comment by diegoweb** [comment]
> so your username format is `local-domain.tld` not `local@domain.tld`?
Only in roundcube I can see this weird format `local-domain.tld`
This is what look like in Virtualmin:
![image](https://user-images.githubusercontent.com/12632119/87823352-818eab00-c849-11ea-89cc-1260048e5b59.png)

In /var/log/maillog the format is `local@domain.tld`

> Or may be you have some other plugin which is messing with the username?
Even with no plugins installed (fresh install) I used to have problems with Password plugin and Virtualmin driver. If I remember well, the old virtualmin driver used to ask in the config file the correct format, so in the driver it would do the str_replace to the correct format. But now this driver has been updated and this is no longer applied.
I've disabled all plugins in Roundcube config.inc.php file, but these 2:

```
	'virtuser_file',
	'markasjunk',
```
I still get this weird format: `local-domain.tld`
May I assume it's not a plugin issue?

> how did the `@` change to a `-`?
I have no idea. Is there a way to debug this? Perhaps write all log in session to a file so we can check when `@` becomes `-`

> Also just for my own curiosity, you have a shared mail system right?
Yes.

> so what is so important about the spamtrap@ address, could you not send the emails to a central address which work for all domains?
From what I understood, Virtualmin Virtual Server uses each domain spamtrap/hamtrap aliases to set the learning stuff for each usermail/domain it's own config. If I had to use a central address, this mean if a user from DOMAIN1 report a newsletter as an SPAM, the DOMAIN2 would be affected.
Maybe am I wrong? I don't know how SPAMASSASSIN works, if it uses the user spamtrap/hamtrap aliases or if it uses the email in attachment and check the headers inside of it. I can ask that in Webmin/Virtualmin support if necessary, since I have a PRO account in there. Let me know if it's needed.

**Comment by johndoh** [comment]
The purpose of the `virtuser_file` plugin is to map between email address and IMAP user but you said that all your IMAP users are email addresses. I guess this is where the issue is coming from. What is in `/etc/postfix/virtual`, Do you really need this plugin enabled?

**Comment by diegoweb** [comment]
God damn it. ROFL.
I really thought that since this plugin is enabled by default, this was a "base plugin" required to Roundcube to work.
It never crossed my mind that I could disable it and test. LOL.

That did the trick, thanks @johndoh.
